File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/bruteforce/Problem13460.java
package com.keunwon.algorithm.bruteforce;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.StringTokenizer;

public class Problem13460 {
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static final int[][] moves = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

    private static int N, M;
    private static char[][] board;

    private static Ball startBall;
    private static boolean[][][][] visited;

    public static void main(String[] args) throws IOException {
        input();
        bfs();
    }

    private static void bfs() {
        var queue = new LinkedList<Ball>();

        queue.offer(startBall);
        visited[startBall.redX][startBall.redY][startBall.blueX][startBall.blueY] = true;

        while (!queue.isEmpty()) {
            var ball = queue.poll();

            if (10 <= ball.count) {
                System.out.println(-1);
                return;
            }

            for (int[] move : moves) {
                var bx = ball.blueX;
                var by = ball.blueY;
                var isBlueHome = false;

                while (board[bx + move[0]][by + move[1]] != '#') {
                    bx += move[0];
                    by += move[1];

                    if (board[bx][by] == 'O') {
                        isBlueHome = true;
                        break;
                    }
                }
                if (isBlueHome) { continue; }

                var rx = ball.redX;
                var ry = ball.redY;
                var isRedHole = false;

                while (board[rx + move[0]][ry + move[1]] != '#') {
                    rx += move[0];
                    ry += move[1];

                    if (board[rx][ry] == 'O') {
                        isRedHole = true;
                        break;
                    }
                }
                if (isRedHole) {
                    System.out.println(ball.count + 1);
                    return;
                }

                if (rx == bx && ry == by) {
                    if (move[0] == -1 && move[1] == 0) {
                        // 상
                        if (ball.redX > ball.blueX) { ++rx; }
                        else { ++bx; }
                    } else if (move[0] == 0 && move[1] == 1) {
                        // 우
                        if (ball.redY > ball.blueY) { --by; }
                        else { --ry; }
                    } else if (move[0] == 1 && move[1] == 0) {
                        // 하
                        if (ball.redX > ball.blueX) { --bx; }
                        else { --rx; }
                    } else if (move[0] == 0 && move[1] == -1) {
                        // 좌
                        if (ball.redY > ball.blueY) { ++ry; }
                        else { ++by; }
                    }
                }

                if (!visited[rx][ry][bx][by]) {
                    visited[rx][ry][bx][by] = true;
                    queue.offer(new Ball(rx, ry, bx, by, ball.count + 1));
                }
            }
        }
        System.out.println(-1);
    }

    private static boolean validMap(int x, int y) {
        return 0 <= x && x < N && 0 <= y && y < M;
    }

    private static void input() throws IOException {
        var st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        board = new char[N][M];
        visited = new boolean[N][M][N][M];

        int rx= 0, ry =0, bx =0, by =0;
        for (int i = 0; i < N; i++) {
            var line = br.readLine();

            for (int j = 0; j < M; j++) {
                var w = line.charAt(j);
                board[i][j] = w;

                if ('R' == w) {
                    rx = i;
                    ry = j;
                } else if ('B' == w) {
                    bx = i;
                    by = j;
                }
            }
        }
        startBall = new Ball(rx, ry, bx, by, 0);
    }

    private static class Ball {
        final int redX;
        final int redY;
        final int blueX;
        final int blueY;
        final int count;

        Ball(int redX, int redY, int blueX, int blueY, int count) {
            this.redX = redX;
            this.redY = redY;
            this.blueX = blueX;
            this.blueY = blueY;
            this.count = count;
        }
    }
}

File: algorithm/kotlin-algorithm/baekjoon/src/main/kotlin/algorithm/baekjoon/Problem13460.kt
package algorithm.baekjoon

import java.util.*

/**
 * <p>
 * 이름: 구슬 탈출 2
 * 난이도: 골드-1
 * </p>
 *
 * @see <a href="https://www.acmicpc.net/problem/13460">구슬 탈출 2 (골드-1)</a>
 **/
class Problem13460 {
    fun solution(n: Int, m: Int, board: Array<CharArray>): Int {
        val visited = Array(n) { Array(m) { Array(n) { BooleanArray(m) } } }
        var (rx, ry, bx, by) = IntArray(4)

        for (i in board.indices) {
            for ((j, type) in board[i].withIndex()) {
                if (type == 'R') {
                    rx = i
                    ry = j
                } else if (type == 'B') {
                    bx = i
                    by = j
                }
            }
        }

        val q = LinkedList<Ball>().apply { offer(Ball(rx, ry, bx, by, 0)) }
        val moves = arrayOf(-1 to 0, 0 to 1, 1 to 0, 0 to -1)
        visited[rx][ry][bx][by] = true

        while (q.isNotEmpty()) {
            val cur = q.poll()

            if (cur.count >= 10) return -1

            for (dir in moves.indices) {
                val (mx, my) = moves[dir]
                var blueX = cur.bx
                var blueY = cur.by
                var isBlueHole = false

                while (board[blueX + mx][blueY + my] != '#') {
                    blueX += mx
                    blueY += my

                    if (board[blueX][blueY] == 'O') {
                        isBlueHole = true
                        break
                    }
                }
                if (isBlueHole) continue

                var redX = cur.rx
                var redY = cur.ry
                var isRedHole = false

                while (board[redX + mx][redY + my] != '#') {
                    redX += mx
                    redY += my

                    if (board[redX][redY] == 'O') {
                        isRedHole = true
                        break
                    }
                }
                if (isRedHole) return cur.count + 1

                if (blueX == redX && blueY == redY) {
                    when (dir) {
                        0 -> if (cur.rx > cur.bx) ++redX else ++blueX
                        1 -> if (cur.ry > cur.by) --blueY else --redY
                        2 -> if (cur.rx > cur.bx) --blueX else --redX
                        3 -> if (cur.ry > cur.by) ++redY else ++blueY
                    }
                }

                if (!visited[redX][redY][blueX][blueY]) {
                    visited[redX][redY][blueX][blueY] = true
                    q.offer(Ball(redX, redY, blueX, blueY, cur.count + 1))
                }
            }
        }

        return -1
    }

    private class Ball(
        val rx: Int,
        val ry: Int,
        val bx: Int,
        val by: Int,
        val count: Int,
    )
}

fun main() {
    val (n, m) = readln().split(" ").map { it.toInt() }
    val board = Array(n) { readln().toCharArray() }

    println(Problem13460().solution(n, m, board))
}
